# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/nsAccessibilityService.cpp
# Commit: d7616232740d
# Full Hash: d7616232740d98904aa4ef6fc6940308faec0846
# Author: Nathan LaPre <nlapre@mozilla.com>
# Date: 2023-06-08 09:27:30
# Regressor Bug: 1789235
# File Overlap Count: 1
# Description:
#   Bug 1789235: Fire a scrolling start event for anchor jump if focus in doc, r=Jamie
#   
#   This revision modifies NotifyOfAnchorJump in order to ensure that we fire a
#   scrolling start event for AT clients. Without this, clients might miss anchor
#   jump updates, since anchor jumps can arrive after getting focus. This revision
# ==============================================================================

diff -r 8d2b1e733b50 -r d7616232740d accessible/base/nsAccessibilityService.cpp
--- a/accessible/base/nsAccessibilityService.cpp	Wed Jun 07 23:24:22 2023 +0000
+++ b/accessible/base/nsAccessibilityService.cpp	Thu Jun 08 00:06:39 2023 +0000
@@ -486,8 +486,24 @@
 void nsAccessibilityService::NotifyOfAnchorJumpTo(nsIContent* aTargetNode) {
   Document* documentNode = aTargetNode->GetUncomposedDoc();
   if (documentNode) {
+    // If the document has focus when we get this notification, ensure that
+    // we fire a start scrolling event.
     DocAccessible* document = GetDocAccessible(documentNode);
-    if (document) document->SetAnchorJump(aTargetNode);
+    const Accessible* focusedAcc = FocusedAccessible();
+    if (focusedAcc && focusedAcc == document) {
+      LocalAccessible* targetAcc = document->GetAccessible(aTargetNode);
+      if (targetAcc) {
+        nsEventShell::FireEvent(nsIAccessibleEvent::EVENT_SCROLLING_START,
+                                targetAcc);
+        document->SetAnchorJump(nullptr);
+      } else {
+        // We can't find the target accessible in the document yet. Set the
+        // anchor jump so that we can fire the scrolling start event later.
+        document->SetAnchorJump(aTargetNode);
+      }
+    } else if (document) {
+      document->SetAnchorJump(aTargetNode);
+    }
   }
 }
 